provider/balance_checker.go

```diff
@@ -2,153 +2,316 @@ package provider
+	leaseCheckCh := make(chan leaseCheckResponse, 1)
+	resultCh := make(chan runner.Result, 1)
+
+	subscriber, err := bc.bus.Subscribe()
+
+	startCh <- err
+	if err != nil {
+		return
+	}
 
 loop:
 	for {
-		withdrawAllNow := false
-
 		select {
-
-		case err := <-bc.lc.ShutdownRequest():
+		case <-bc.lc.ShutdownRequest():
 			bc.log.Debug("shutting down")
-			bc.lc.ShutdownInitiated(err)
+			bc.lc.ShutdownInitiated(nil)
 			break loop
-		case <-tick.C:
-			tick.Stop() // Stop the timer
-			// Start the balance check
-			balanceCheckResult = runner.Do(func() runner.Result {
-				return runner.NewResult(bc.doCheck(ctx))
-			})
-
-		case balanceCheck := <-balanceCheckResult:
-			balanceCheckResult = nil
-			tick.Reset(bc.cfg.PollingPeriod) // Re-enable the timer
-			err := balanceCheck.Error()
-			if err != nil {
-				bc.log.Error("failed to check balance", "err", err)
-				break
-			}
+		case evt := <-subscriber.Events():
+			switch ev := evt.(type) {
+			case event.LeaseAddFundsMonitor:
+				var scheduledWithdraw time.Time
+				// if provider configured with periodic force withdrawal
+				// set next time at which withdraw will happen
+				if bc.cfg.WithdrawalPeriod > 0 {
+					scheduledWithdraw = time.Now().Add(bc.cfg.WithdrawalPeriod)
+				}
+
+				lState := &leaseState{
+					scheduledWithdrawAt: scheduledWithdraw,
+				}
+
+				bc.leases[ev.LeaseID] = lState
 
-			tooLow := balanceCheck.Value().(bool)
-			if tooLow {
-				// trigger the withdrawal
-				bc.log.Info("balance below target amount")
-				withdrawAllNow = true
+				// if there was provider restart with bunch of active leases
+				// spread their requests across 1min interval
+				// to reduce pressure on the RPC
+				if !ev.IsNewLease {
+					checkIn := time.Duration(rand.Int63n(int64(time.Minute))) // nolint: gosec
+					lState.tm = bc.timerFunc(ctx, checkIn, ev.LeaseID, false, leaseCheckCh)
+				} else {
+					bc.runEscrowCheck(ctx, ev.LeaseID, false, leaseCheckCh)
+				}
+			case event.LeaseRemoveFundsMonitor:
+				lsState, exists := bc.leases[ev.LeaseID]
+				if !exists {
+					break
+				}
+
+				if lsState.tm != nil && !lsState.tm.Stop() {
+					<-lsState.tm.C
+				}
+
+				delete(bc.leases, ev.LeaseID)
 			}
-		case withdrawAll := <-withdrawAllResult:
+		case res := <-leaseCheckCh:
+			// we may have timer fired just a heart beat ahead of lease remove event.
+			lState, exists := bc.leases[res.lid]
+			if !exists {
+				continue loop
+			}
+
+			withdraw := false
 
-			withdrawAllResult = nil
-			withdrawalTicker.Reset(bc.cfg.PollingPeriod) // Re-enable the timer
-			if err := withdrawAll.Error(); err != nil {
-				bc.log.Error("failed to started withdrawals", "err", err)
+			switch res.state {
+			case respStateScheduledWithdraw:
+				// reschedule periodic withdraw if configured
+				if bc.cfg.WithdrawalPeriod > 0 {
+					lState.scheduledWithdrawAt = time.Now().Add(bc.cfg.WithdrawalPeriod)
+				}
+				fallthrough
+			case respStateOutOfFunds:
```

Comment:
My understanding is this should only happen once per lease, so we should probably log this at an info level? It's significant enough I would think.

Reply:
That's actually a good call. If the lease is running for a year provider won't be able to access funds until withdrawn.